### PR TITLE
Adding key to exclude target namespace in operatorgroup

### PR DESCRIPTION
**Pull request by claudiol**
Implements the issue describe in [Issue #296 ](https://github.com/validatedpatterns/common/issues/296)

- Added operatorgroupExcludeTargetNS key to exclude targetNamespace in an Operator Group

### CODE REVIEW

File: clustergroup/values.schema.json [hunk 1]
```diff
@@ -266,6 +266,13 @@
             "type": "string"
           }
         },
+        "operatorgroupExcludeTargetNS": {
+          "type": "array",
+          "description": "List of namespaces to exclude the target namespace.",
```
Comment (mbaldessari) [inline clustergroup/values.schema.json:271]: Maybe call it `Specify the list of namespaces where the target namespace field in the corresponding operatorgroup object should be excluded`

**Review by mbaldessari — COMMENTED** [review]
Can you also add a small example in examples/values-example.yaml?

I am not entirely sure about the API/naming used. On one side there is precedent already via the operatorgroupExcludes and on the other hand I am not super fond of adding yet another variable. Maybe we should consider adding a field in the namespace list, like we do for annotations and labels? (I do not love this one either, maybe it's because it is Monday :) )

File: clustergroup/values.schema.json (hunk 1 above)
Comment (claudiol) [reply clustergroup/values.schema.json]: Done.

**Comment by claudiol** [comment]
@mbaldessari Potential alternative if we want to exclude target namespace from OperatorGroup:

```
  namespaces:
  - open-cluster-management:
      labels:
        openshift.io/node-selector: ""
        kubernetes.io/os: linux
      annotations:
        openshift.io/cluster-monitoring: "true"
        owner: "namespace owner"
      excludeOperatorGroupTargetNS: true
  - application-ci:
      excludeOperatorGroupTargetNS: true
  - include-ci
  - excludes-ci

  operatorgroupExcludes:
  - excludes-ci
  ```

**Comment by claudiol** [comment]
> @mbaldessari Potential alternative if we want to exclude target namespace from OperatorGroup:
> 
> ```
>   namespaces:
>   - open-cluster-management:
>       labels:
>         openshift.io/node-selector: ""
>         kubernetes.io/os: linux
>       annotations:
>         openshift.io/cluster-monitoring: "true"
>         owner: "namespace owner"
>       excludeOperatorGroupTargetNS: true
>   - application-ci:
>       excludeOperatorGroupTargetNS: true
>   - include-ci
>   - excludes-ci
> 
>   operatorgroupExcludes:
>   - excludes-ci
> ```
Here's a updated approach to include operatorgroup definitions inside the namespaces construct.
- The idea is to eventually deprecate **operatorgroupExcludes**
- Add the **operatorgroup** to the definition of namespaces in our values files when defined as a map

The construct would look something like this:
1. Define the namespace as a map and explicitly exclude the **namespace** from  **targetNamespaces** in the OperatorGroup manifest
```
   namespaces:
   - application-ci:
     labels:
        openshift.io/node-selector: ""
        kubernetes.io/os: linux
     annotations:
        openshift.io/cluster-monitoring: "true"
        owner: "namespace owner"
     operatorgroup:
        targetNamespaces:
```
2. Define the namespace as a map and include a list of **targetNamespaces** in the OperatorGroup manifest
```
   namespaces:
   - application-ci:
     labels:
        openshift.io/node-selector: ""
        kubernetes.io/os: linux
     annotations:
        openshift.io/cluster-monitoring: "true"
        owner: "namespace owner"
     operatorgroup:
        targetNamespaces:
          - application-ci
          - namespace2
```
``NOTE: You must explicitly list your own namespace. It will not included by the framework in the generated manifest``
3. Define the namespace as a map and not including the **operatorgroup** key. By default this will add the **namespace** to **targetNamespaces**
```
   namespaces:
   - application-ci:
     labels:
        openshift.io/node-selector: ""
        kubernetes.io/os: linux
     annotations:
        openshift.io/cluster-monitoring: "true"
        owner: "namespace owner"
```
4. If the namespace is defined as a list the default is to include the **namespace** in  **targetNamespaces**
``` 
  namespaces:
  - application-ci
```
5. To prevent the framework to create an OperatorGroup for the namespace we will still include the namespace in the **operatorgroupExcludes** list.

File: examples/values-example.yaml [hunk 2]
```diff
@@ -23,13 +23,20 @@ clusterGroup:
       annotations:
         openshift.io/cluster-monitoring: "true"
         owner: "namespace owner"
-  - application-ci
+  - application-ci:
+      operatorgroup:
+        targetNamespaces:
+          - application-ci
+          - other-namespace
```
Comment (beekhof) [inline examples/values-example.yaml:30]: I would suggest decoupling `operatorgroup` and `targetNamespaces`
Also, be consistent with camelCase.

```suggestion
      operatorGroup: true # true being the default anyway
      targetNamespaces:
        - application-ci
        - other-namespace
```

File: examples/values-example.yaml [hunk 3]
```diff
@@ -23,13 +23,20 @@ clusterGroup:
...
+  - exclude-targetns:
+      operatorgroup:
+        targetNamespaces:
+  - include-ci
   - excludes-ci
```
Comment (beekhof) [inline examples/values-example.yaml:35]:
```suggestion
  - excludes-ci
     operatorGroup: false
```

File: examples/values-example.yaml (hunk 2 above)
Comment (claudiol) [reply examples/values-example.yaml]: Thanks for the suggestion and it totally makes sense.  Updated code to support this.

File: examples/values-example.yaml (hunk 3 above)
Comment (claudiol) [reply examples/values-example.yaml]: Updated values-example.yaml with change.

**Comment by claudiol** [comment]
> > @mbaldessari Potential alternative if we want to exclude target namespace from OperatorGroup:
> > ```
> >   namespaces:
> >   - open-cluster-management:
> >       labels:
> >         openshift.io/node-selector: ""
> >         kubernetes.io/os: linux
> >       annotations:
> >         openshift.io/cluster-monitoring: "true"
> >         owner: "namespace owner"
> >       excludeOperatorGroupTargetNS: true
> >   - application-ci:
> >       excludeOperatorGroupTargetNS: true
> >   - include-ci
> >   - excludes-ci
> > 
> >   operatorgroupExcludes:
> >   - excludes-ci
> > ```
> 
> Here's a updated approach to include operatorgroup definitions inside the namespaces construct.
> 
> * The idea is to eventually deprecate **operatorgroupExcludes**
> * Add the **operatorgroup** to the definition of namespaces in our values files when defined as a map
> 
> The construct would look something like this:
> 
> 1. Define the namespace as a map and explicitly exclude the **namespace** from  **targetNamespaces** in the OperatorGroup manifest
> 
> ```
>    namespaces:
>    - application-ci:
>      labels:
>         openshift.io/node-selector: ""
>         kubernetes.io/os: linux
>      annotations:
>         openshift.io/cluster-monitoring: "true"
>         owner: "namespace owner"
>      operatorgroup:
>         targetNamespaces:
> ```
> 
> 2. Define the namespace as a map and include a list of **targetNamespaces** in the OperatorGroup manifest
> 
> ```
>    namespaces:
>    - application-ci:
>      labels:
>         openshift.io/node-selector: ""
>         kubernetes.io/os: linux
>      annotations:
>         openshift.io/cluster-monitoring: "true"
>         owner: "namespace owner"
>      operatorgroup:
>         targetNamespaces:
>           - application-ci
>           - namespace2
> ```
> 
> `NOTE: You must explicitly list your own namespace. It will not included by the framework in the generated manifest` 3. Define the namespace as a map and not including the **operatorgroup** key. By default this will add the **namespace** to **targetNamespaces**
> 
> ```
>    namespaces:
>    - application-ci:
>      labels:
>         openshift.io/node-selector: ""
>         kubernetes.io/os: linux
>      annotations:
>         openshift.io/cluster-monitoring: "true"
>         owner: "namespace owner"
> ```
> 
> 4. If the namespace is defined as a list the default is to include the **namespace** in  **targetNamespaces**
> 
> ```
>   namespaces:
>   - application-ci
> ```
> 
> 5. To prevent the framework to create an OperatorGroup for the namespace we will still include the namespace in the **operatorgroupExcludes** list.

We now use the following namespace structure to include/exclude an OperatorGroup:

1. Create OperatorGroup for the Namespace and exclude targetNamespaces from the OperatorGroup definition
```
   namespaces:
   - application-ci:
     labels:
        openshift.io/node-selector: ""
        kubernetes.io/os: linux
     annotations:
        openshift.io/cluster-monitoring: "true"
        owner: "namespace owner"
     operatorGroup: true
     targetNamespaces:
```
2. Do not create the OperatorGroup for the Namespace
```
   namespaces:
   - exclude-operatorgroup:
     operatorGroup: false
```
3. Create the OperatorGroup for the Namespace and add the namespace to the targetNamespaces
```
   namespaces:
   - include-targetns:
     operatorGroup: true
```
Same result if you just have it as a list and not a map
```
   namespaces:
   - include-targetns
```

File: clustergroup/templates/core/operatorgroup.yaml [hunk 4]
```diff
@@ -5,17 +5,23 @@
...
-
+  {{- if $v.operatorGroup }}{{- /* Checks if the user sets operatorGroup: false */}}
 apiVersion: operators.coreos.com/v1
 kind: OperatorGroup
 metadata:
   name: {{ $k }}-operator-group
   namespace: {{ $k }}
 spec:
   targetNamespaces:
+    {{- if or (and $v.operatorGroup (hasKey $v "targetNamespaces")) (and $v.operatorGroup (not (empty $v.targetNamespaces))) }}
```
Comment (beekhof) [inline clustergroup/templates/core/operatorgroup.yaml:16]:
```suggestion
    {{- if (hasKey $v "targetNamespaces") }}
```

**Pull request merged**